File: algo/datastruct/tree/suffix-tree/src/stree.py
#!/usr/bin/python

# strie.py, Suffix Trie.
# Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import string

class Node:
    def __init__(self, suffix=None):
        self.children = {} # 'c':(word, Node), where word = (l, r)
        self.suffix = suffix

class STree:
    def __init__(self, s):
        self.str = s
        self.infinity = len(s)+1000
        self.root = Node()

    def substr(self, sref):
        return substr(self.str, sref)

#
# In order to improve the efficiency by operation on-line
# There is only one copy of the string. All sub strings are
# represented as reference pair: 
#   w = (k, p)
# where: w = str[k:p+1], +1 is because of Python's specific
# problem
#
def substr(str, str_ref):
    (l, r)=str_ref
    return str[l:r+1]

def length(str_ref):
    (l, r)=str_ref
    return r-l+1

#
# Algorithm 2 in Ukkonen '95
#   l: left index of a word, k in Ukkonen '95
#   node: s in Ukkonen '95
#   None: _|_ in Ukkonen '95
#
def suffix_tree(str):
    t = STree(str)
    node = t.root # init active point is (root, Empty)
    l = 0
    for i in range(len(str)):
        (node, l) = update(t, node, (l, i))
        (node, l) = canonize(t, node, (l, i))
    return t

#
# Main func: STree(Str[i-1]) ==> STree(Str[i])
#   param: (node, (l, i-1)): AP (Active Point)
#   prev: oldr in Ukkonen '95
#   p:    r in Ukkonen '95
#   return, EP (End Point) ref pair (node, (l, i-1))
#
def update(t, node, str_ref):
    (l, i) = str_ref 
    c = t.str[i] # current char
    prev = Node() # dummy init 
    while True:
        (finish, p) = branch(t, node, (l, i-1), c)
        if finish:
            break
        p.children[c]=((i, t.infinity), Node())
        prev.suffix = p
        prev = p
        # go up along suffix link
        (node, l) = canonize(t, node.suffix, (l, i-1))
    prev.suffix = node
    return (node, l)

#
# branch: 
#  this is test-and-split function in Ukkonen '95
#  test if (node, str_ref) is an EP, (EP has a c-transition)
#  if not, then branch out a new node
#
def branch(t, node, str_ref, c):
    (l, r) = str_ref
    if length(str_ref)<=0: # (node, empty)
        if node is None: #_|_
            return (True, t.root)
        else:
            return ((c in node.children), node)
    else:
        ((l1, r1), node1) = node.children[t.str[l]]
        pos = l1+length(str_ref)
        if t.str[pos]==c:
            return (True, node)
        else:             # node--->branch_node--->node1
            branch_node = Node()
            node.children[t.str[l1]]=((l1, pos-1), branch_node)
            branch_node.children[t.str[pos]] = ((pos, r1), node1)
            return (False, branch_node)

#
# node[c]--->(l, r), _ 
# node[c]--->((l', r'), node')--->...-->((l'', r''), node'')--->((x, inf), leaf)
# where _: it may not be a node, but some implicity position
# find the closet node and left, so that they point to same position _
# special case: ||None, (k, p)|| = (root, (k+1, p))
#
def canonize(t, node, str_ref):
    (l, r) = str_ref 
    if node is None:
        if length(str_ref)<=0:
            return (None, l)
        else:
            return canonize(t, t.root, (l+1, r))
    while l<=r: # str_ref is not empty
        ((l1, r1), child) = node.children[t.str[l]] # node--(l', r')-->child
        if r-l >= r1-l1: #node--(l',r')-->child--->...
            l += r1-l1+1 # remove |(l',r')| chars from (l, r)
            node = child 
        else:
            break
    return (node, l)

def to_lines(t, node):
    if len(node.children)==0:
        return [""]
    res = []
    for c, (str_ref, tr) in sorted(node.children.items()):
        lines = to_lines(t, tr)
        edge_str = substr(t.str, str_ref)
        lines[0] = "|--"+edge_str+"-->"+lines[0]
        if len(node.children)>1:
            lines[1:] = map(lambda l: "|"+" "*(len(edge_str)+5)+l, lines[1:])
        else:
            lines[1:] = map(lambda l: " "+" "*(len(edge_str)+6)+l, lines[1:])
        if res !=[]:
            res.append("|")
        res += lines
    return res

def to_str(t):
    return "\n".join(to_lines(t, t.root))

class SuffixTreeTest:
    def __init__(self):
        print "start suffix tree test"

    def run(self):
        strs = ["cacao", "mississippi", "banana$"]
        for s in strs:
            self.test_build(s)

    def test_build(self, str):
        for i in range(len(str)):
            self.__test_build(str[:i+1])


    def __test_build(self, str):
        print "Suffix Tree ("+str+"):\n", to_str(suffix_tree(str)),"\n"

if __name__ == "__main__":
    SuffixTreeTest().run()

File: algo/datastruct/tree/suffix-tree/src/STree.hs
{-
    Patricia.hs, Alphabetic Patricia Tree.
    Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-}

module STree where

import Data.List

-- method 1
-- According to:
--   * Robert Giegerich and Stefan Kurtz, \"/A comparison of
--     imperative and purely functional suffix tree constructions/\",
--     Science of Computer Programming 25(2-3):187-218, 1995,
--     <http://citeseer.ist.psu.edu/giegerich95comparison.html>
--
-- Liu Xinyu: This method is a kind of "Brute Force" algorithm in fact.
--  It can't match O(n), n=length(word), as Ukkonen, McCreight and Weiner's.
--  Altough it can benifit from Haskell's lazy evaluation, so that
--  without traverse, the tree won't be constructed.
--
--  However, in other functional programming language, such as ML, Scheme/Lisp
--  Which doesn't do lazy evaulation by default, this method
--  is equivalent to:
--      foldl patricia-insert empty tails(word)
--
data STree = Leaf | Branch [(Edge, STree)] deriving (Eq, Show)
type Edge = (String, Int)
type EdgeFunction = [String]->(Int, [String])

alpha = ['a'..'z']++['A'..'Z']

lazyTree::EdgeFunction->[String]->STree
lazyTree edge = stree where
    stree [[]] = Leaf
    stree ss = Branch [((a:sa, 1+cpl), stree ss') 
                           | a<-alpha,
                             x@(sa:_) <- [ss `selectBy` a],
                             (cpl, ss') <- [edge x]]
    selectBy ss a = [cs | c:cs <-ss, c==a] -- select suffixes starting with a

--ast: Atomic Suffix Tree,  means Suffix Trie
ast::EdgeFunction
ast ss = (0, ss)

--cst: Compact Suffix Tree, Suffix Patricia
--     extract the longest common prefix
cst::EdgeFunction
cst [s] = (length s, [[]])
cst awss@((a:w):ss) | null [c | c:_<-ss, a /= c] = (1+cpl, ss')
                    | otherwise                  = (0, awss)
                    where (cpl, ss') = cst (w:[u | _:u<-ss])

suffixes = init.tails -- returns non-empty suffixes

suffixTrie::String->STree
suffixTrie = lazyTree ast . suffixes

suffixTree::String->STree
suffixTree = lazyTree cst . suffixes

-- method 2
-- Liu Xinyu: pure brute-force method with list comprehension
--
data Tr = Lf | Br [(String, Tr)] deriving (Eq, Show)
type EdgeFunc = [String]->(String, [String])

-- Create Lazy Radix Tree
-- some helper comments
--  xs@(x:_)<-[...], The @(x:_) is used to filter out the empty elements
lazyTr::EdgeFunc -> [String] -> Tr
lazyTr edge = build where
    build [[]] = Lf
    build ss = Br [(a:prefix, build ss') | a<-alpha, 
                                           xs@(x:_) <-[[cs | c:cs<-ss, c==a]],
                                           (prefix, ss')<-[edge xs]]

-- Trie
edgeTrie::EdgeFunc
edgeTrie ss = ("", ss)

-- Patricia: Extract the longest common prefix
-- ex: 
--   edgeTree ["an", "another", "and"] = ("an", ["", "other", "d"])
--   edgeTree ["bool", "foo", "bar"] = ("", ["bool", "foo", "bar"])
--
-- some helper comments
--   let awss@((a:w):ss) = ["an", "another", "and"]
--       (a:w) = "an",  ss = ["another", "and"]
--       a='a', w="n"
--       rests awss = w:[u| _:u<-ss] = ["n", "nother", "nd"]
--
edgeTree::EdgeFunc
edgeTree [s] = (s, [[]])
edgeTree awss@((a:w):ss) | null [c|c:_<-ss, a/=c] = (a:prefix, ss')
                         | otherwise              = ("", awss)
                         where (prefix, ss') = edgeTree (w:[u| _:u<-ss])
edgeTree ss = ("", ss) -- (a:w):ss can't be match <==> head ss == ""

sufTrie::String->Tr
sufTrie = lazyTr edgeTrie . tails

sufTree::String->Tr
sufTree = lazyTr edgeTree . tails

-- Let's prove it's brute-force
trie::[String]->Tr
trie = lazyTr edgeTrie

patricia::[String]->Tr
patricia = lazyTr edgeTree

-- Method 3
-- Liu Xinyu: Construct the suffix tree from left to right
update::Tr->Char->Tr
update Lf c = Br [([c], Lf)]
update (Br ts) c = case find (\(s, _)->c==head s) ts of 
                   Nothing -> Br ((map (insertAfter c c') ts)++[([c], Lf)])
                   _       -> Br ((map (insertAfter c c') ts)) --Br (map (appendLeaf c) ts)
    where
      c' = lastElem $ head ts -- c' is the last char of any edge to leaf
                                      
appendLeaf::Char->(String, Tr)->(String, Tr)
appendLeaf c (s, Lf) = (s++[c], Lf)
appendLeaf c (s, (Br ts)) = (s, Br (map (appendLeaf c) ts))
                   
lastElem::(String, Tr)->Char
lastElem (s, Lf) = last s
lastElem (_, Br ts) = lastElem $ head ts

--(s1c's2c'...snc's, t) --> (s1c', Br [(c, Lf), (s2c', Br [(c, Lf), Br ...
insertAfter::Char->Char->(String, Tr)->(String, Tr)
insertAfter c c' (s, t) = ins c lst where
	lst = split (c'==) s
        ins c [x] = if isLeaf t then (x++[c], t) else (x, update t c)
	ins c (x:xs) = (x, Br [(ins c xs), ([c], Lf)])

isLeaf::Tr -> Bool
isLeaf Lf = True
isLeaf _  = False

split _ [] = []
split f lst | ys == [] = [xs] 
            | otherwise = (xs++[head ys]): split f (tail ys)
    where (xs, ys) = break f lst

suffixTree'::String->Tr
suffixTree' = foldl update Lf

-- testing

testSuffixTrie s = "Robert, Kurtz:\n SuffixTrie(\"" ++ s ++ "\")=" ++ (show $ suffixTrie s) ++ "\n"

testSuffixTree s = "Robert, Kurtz:\n SuffixTree(\"" ++ s ++ "\")=" ++ (show $ suffixTree s) ++ "\n"

testSuffixTrie' s = "Liu Xinyu:\n SuffixTrie(\"" ++ s ++ "\")=" ++ (show $ sufTrie s) ++ "\n"

testSuffixTree' s = "Liu Xinyu:\n SuffixTree(\"" ++ s ++ "\")=" ++ (show $ sufTree s) ++ "\n"

testTrie ss = "Trie(\"" ++ (show ss) ++ "\")=" ++ (show $ trie ss) ++ "\n"

testPatricia ss = "Patricia(\"" ++ (show ss) ++ "\")=" ++ (show $ patricia ss) ++ "\n"

test = concat [ f s | s <- ["cacao", "mississippi", "banans"],
                      f <- [testSuffixTrie, testSuffixTree, testSuffixTrie', testSuffixTree']] ++
       testTrie ["zoo", "bool", "boy", "another", "an", "a"] ++
       testPatricia ["zoo", "bool", "boy", "another", "an", "a"]